websocketOutportDriver/client/wsClient.go

```diff
@@ -0,0 +1,188 @@
+	RetryDurationInSec       uint32
+	BlockingAckOnError       bool
+	OperationHandler         OperationHandler
+	PayloadParser            PayloadParser
+	Uint64ByteSliceConverter sender.Uint64ByteSliceConverter
+	WSConnClient             WSConnClient
+}
+
+// NewWsClientHandler will create a ws client to receive data from an observer/light client
+func NewWsClientHandler(args *ArgsWsClient) (*client, error) {
+	if args.OperationHandler == nil {
+		return nil, errNilOperationHandler
+	}
+	if args.PayloadParser == nil {
+		return nil, errNilPayloadParser
+	}
+	if args.WSConnClient == nil {
+		return nil, errNilWsConnReceiver
+	}
+	if args.Uint64ByteSliceConverter == nil {
+		return nil, errNilUint64ByteSliceConverter
+	}
+	if len(args.Url) == 0 {
+		return nil, errEmptyUrlProvided
+	}
+
+	urlReceiveData := url.URL{Scheme: "ws", Host: args.Url, Path: data.WSRoute}
+	retryDuration := time.Duration(args.RetryDurationInSec) * time.Second
+
+	return &client{
+		url:                      urlReceiveData.String(),
+		blockingOnError:          args.BlockingAckOnError,
+		retryDuration:            retryDuration,
+		wsConn:                   args.WSConnClient,
+		payloadParser:            args.PayloadParser,
+		operationHandler:         args.OperationHandler,
+		uint64ByteSliceConverter: args.Uint64ByteSliceConverter,
+	}, nil
+}
+
+// Start will start the client listening ws process
+func (c *client) Start() {
+	log.Info("connecting to", "url", c.url)
+
+	for {
+		err := c.wsConn.OpenConnection(c.url)
+		if err != nil {
+			log.Warn(fmt.Sprintf("c.openConnection(), retrying in %v...", c.retryDuration), "error", err.Error())
+			time.Sleep(c.retryDuration)
+			continue
+		}
+
+		closed := c.listenOnWebSocket()
+		if closed {
+			return
+		}
+	}
+}
+
+func (c *client) listenOnWebSocket() (closed bool) {
+	for {
+		_, message, err := c.wsConn.ReadMessage()
+		if err == nil {
+			c.verifyPayloadAndSendAckIfNeeded(message)
+			continue
+		}
+
+		_, isConnectionClosed := err.(*websocket.CloseError)
+		if !isConnectionClosed {
+			if strings.Contains(err.Error(), closedConnection) {
+				return true
+			}
+			log.Warn("c.listenOnWebSocket()-> connection problem, retrying", "error", err.Error())
+		} else {
+			log.Warn(fmt.Sprintf("websocket terminated by the server side, retrying in %v...", c.retryDuration), "error", err.Error())
+		}
+		return
+	}
+}
+
+func (c *client) verifyPayloadAndSendAckIfNeeded(payload []byte) {
+	if len(payload) == 0 {
+		log.Error("empty payload")
+		return
+	}
+
+	payloadData, err := c.payloadParser.ExtractPayloadData(payload)
+	if err != nil {
+		log.Error("error while extracting payload data: " + err.Error())
+		return
+	}
+
+	log.Info("processing payload",
+		"counter", payloadData.Counter,
+		"operation type", payloadData.OperationType,
+		"payload", payloadData.Payload,
+		"message length", len(payloadData.Payload),
+	)
+
+	function, ok := c.operationHandler.GetOperationHandler(payloadData.OperationType)
```

Comment:
this can be cleaned a little bit. Instead of having an interface that contains the GetOperationHandler method that provide a pointer function to be called a few lines later, why not refactor the operationsHandler (maybe call it PayloadProcessor or something like this) that can be described as following:
```GO
type PayloadProcessor interface{
    ProcessPayload(payload *websocketOutportDriver.PayloadData) error
    IsInterfaceNil() bool
}
``` 

the correct function selection (or perhaps implementation that will handle a certain type of payload) will be handled by the above processor implementation.

Reply:
Right, very good suggestion.
I was also thinking about this implementation, but was not sure how much I should refactor this.
Will change it as proposed